protocol_connect.go

```diff
@@ -820,6 +830,45 @@ func (u *connectStreamingUnmarshaler) EndStreamError() *Error {
 	return u.endStreamErr
 }
 
+type connectGetMarshaler struct {
+	getURLMaxBytes int
+	codec          stableCodec
+	duplexCall     *duplexHTTPCall
+	nextMarshaler  connectMarshaler
+}
+
+func (m *connectGetMarshaler) Marshal(message any) *Error {
+	if m.tryGet(message) {
+		return nil
+	}
+	return m.nextMarshaler.Marshal(message)
+}
+
+func (m *connectGetMarshaler) tryGet(message any) bool {
+	data, err := m.codec.MarshalStable(message)
+	if err != nil || len(data) > m.getURLMaxBytes {
+		// TODO(jchadwick-buf): should return error? emit warning? unclear.
+		return false
+	}
+	url := *m.duplexCall.URL()
+	query := url.Query()
+	query.Set(connectUnaryAPIQueryParameter, connectProtocolVersionQueryValue)
+	query.Set(connectUnaryEncodingQueryParameter, m.codec.Name())
+	if m.codec.IsBinary() {
+		query.Set(connectUnaryMessageQueryParameter, EncodeBinaryHeader(data))
+		query.Set(connectUnaryBase64QueryParameter, "1")
+	} else {
+		query.Set(connectUnaryMessageQueryParameter, string(data))
+	}
+	url.RawQuery = query.Encode()
+	if len(url.String()) > m.getURLMaxBytes {
+		return false
+	}
+	m.duplexCall.SetMethod(http.MethodGet)
+	*m.duplexCall.URL() = url
+	return true
+}
```

Comment:
Few thoughts here:

* We're missing support for the `cmp` parameter, which indicates the compression scheme in use. (I think we may have missed that on the server side, too.)
* Ideally, we wouldn't ever marshal the same message twice. When we call `MarshalStable`, we get back a bag of data that ought to be usable for a `POST` request too - it's a valid serialization of the request message. If it's too big for a `GET`, we should try compressing it; if it's still too big, we should hand off the serialized-and-compressed data to the regular `connectMarshaler`. Can we just add a method to `connectMarshaler` to make this work?

Reply:
Unfortunately, accomplishing all of this is essentially impossible with the way marshalers are laid out today. We need to do things that are request-specific for Get, but `connectUnaryMarshaler` is used for both requests and responses. Moreover, the control flow of `connectUnaryMarshaler.Marshal` is subtley different from what we want for HTTP `GET`, so it's hard to extract out the commonality. I wound up mostly duplicating the `connectUnaryMarshaler.Marshal` function. I'm open to suggestions here, but I suspect getting this to be any cleaner will require a deeper change. d270803c63ee6c7961e2c752022e5f656e6eaf1d